README.md
# No Budget

_text-mode user interface for [convergent XEN farms](https://github.com/pbraun9/xen)_

Build and install nobudget onto the system.

	cd nobudget/
	make
	make install

Eventually disable password-based authentication altogether and take over 22/tcp.

_assuming the casual SSH daemon is running on another port_

	vi /etc/ssh/sshd_config_nobudget

	AllowGroups budgetusers
	PermitRootLogin no
	Port 22
	PidFile /var/run/sshd_nobudget.pid
	HostKey /etc/ssh/ssh_host_ed25519_key
	HostKey /etc/ssh/ssh_host_ecdsa_key
	# no sftp

	AuthenticationMethods publickey
	AuthorizedKeysFile .ssh/authorized_keys
	Ciphers aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
	KbdInteractiveAuthentication no
	MaxAuthTries 3
	PasswordAuthentication no
	PermitEmptyPasswords no
	PrintMotd no
	Protocol 2
	StrictModes yes
	UseDNS no
	UsePAM no
	X11Forwarding no
	AllowTcpForwarding no

	vi /etc/rc.d/rc.local

	echo -n starting nobudget sshd ...
	/usr/sbin/sshd -f /etc/ssh/sshd_config_nobudget && echo done || echo FAIL

hard-code the shared group for nobudget users

	groupadd -g 1004 budgetusers

create users and let them reach the text UI

	user=USERNAME

	useradd -m -g budgetusers -b /data/users -s /usr/local/bin/nobudget $user
	chmod 700 /data/users/$user/
	passwd --unlock $user

and put user's SSH public key in place.



CHANGELOG.md

v0.0.1

- cycle back to the main menu (thanks to Artem)

v0.0.0

- shell scripts to handle guest creation and maintenance
- initial ncurses UI (Pierre and Alastar)



nobudget.c
#include "nobudget.h"

int lenght = 0;
int longest = 0;

char title[] = "Definitely Not a Cloud";
char footer_left[] = "pmr.angrycow.ru";
char footer_right[] = "nobudget v0.0.1";
int left_column_length = 28;
struct menu_ui start_menu;
char *choices[] = {
	"N E W   G U E S T",
        "M A N A G E   G U E S T S",
        "S U P P O R T",
	"Q U I T"};

int n_choices = sizeof(choices) / sizeof(char *);

int menu_is_active = 1;

void startwin()
{
	initscr();

	// debug mode allows ^C ^Z
	cbreak();
	//raw();

	keypad(stdscr, TRUE);
	noecho();
	curs_set(0);
	return;
}

int main()
{
	int highlight = 1;
	int choice = 0;
	int c;

	startwin();

	int i;
	for (i = 0; i < n_choices; ++i)
	{
		lenght = strlen(choices[i]);
		if (lenght > longest)
			longest = lenght;
	}

	print_menu(highlight);
	while (menu_is_active)
	{
		c = getch();
		switch (c)
		{
		case KEY_UP:
			if (highlight > 1)
				--highlight;
			break;
		case KEY_DOWN:
			if (highlight < n_choices)
				++highlight;
			break;
		// ascii code for ENTER
		case 10:
		{
			choice = highlight;
			break;
		}
		// ascii code for ESC
		case 27:
		{
			menu_is_active = 0;
			break;
		}
		}
		menu_handler(choice);
		choice = 0;

		print_menu(highlight);
		// if(choice != 0)	/* User did a choice come out of the infinite loop */
		// 	break;
	}
	// restore initial terminal settings
	endwin();
	return 0;
}

void print_menu(int highlight)
{
	int x, y, i;

	erase();

	attron(A_BOLD);
	mvprintw(0, (COLS - strlen(title)) / 2 - 2, "%s", title);
	attroff(A_BOLD);

	x = (COLS - longest) / 2 - 2;
	y = (LINES - n_choices) / 2 - 1 ;
	for (i = 0; i < n_choices; ++i)
	{
		if (highlight == i + 1) /* High light the present choice */
		{
			attron(A_REVERSE);
			mvprintw(y, x, "%s", choices[i]);
			attroff(A_REVERSE);
		}
		else
		{
			mvprintw(y, x, "%s", choices[i]);
		}
		// we want an empty line between the menu entries
		++y;
		++y;
	}

	mvprintw(LINES - 1, 1, "%s", footer_left);
	mvprintw(LINES - 1, COLS - strlen(footer_right) - 1, "%s", footer_right);

	refresh();
}

void menu_handler(int selected_item)
{
	/* Processes user's menu item choice

	Called in:
		nobudget.c : main()
	Args:
		selected_item(int): a number of menu item selected
	Returns:
		None
	*/
	switch (selected_item)
	{
		case 1:
		{
			// NEW GUEST
			endwin();
			system("/usr/local/bin/new-guest.bash");
			startwin();
			break;
		}
		case 2:
		{
			// MANAGE GUESTS
			endwin();
			system("/usr/local/bin/manage-guests.bash");
			startwin();
			break;
		}
		case 3:
		{
			// SUPPORT
			endwin();
			//display_text("/usr/local/lib/support.txt");
			system("/usr/local/bin/display-support.bash");
			startwin();
			break;
		}
		case 4:
		{
			// QUIT
			endwin();
			exit(0);
		}
		default:
		{
			// IDLING
			break;
		}
	}
}

void display_text(char *support_file_path)
{
	/* Displays content of support file on screen

	Called in:
		nobudget.c : menu_handler()
	Args:
		support_file_path(char *): a path to text file
	Returns:
		None
	*/

	// open a file
	FILE *fp;
	char *line = NULL;
	size_t len = 0;
	ssize_t read;
	fp = fopen(support_file_path, "r");

	// handle an exceprion when file cannot be opened
	if (fp == NULL)
	{
		exit(EXIT_FAILURE);
	}

	// clear screen before printing output
	erase();

	// output the text line by line
	int current_line = 1;
	while ((read = getline(&line, &len, fp)) != -1)
	{
		mvprintw(current_line, 1, line);
		++current_line;
	}
	fclose(fp);
}



newguest.bash
#!/bin/bash

debug=0

[[ $1 = auto ]] && auto=1

function asksystem {
	cat <<EOF

 Which system would you like to deploy?

  1. Debian 11 (Bullseye)
  2. NetBSD current (Sep 2022)
  3. Slackware Linux 15.0
  4. Ubuntu 22 (Jammy Jellyfish)

EOF
	  #- Sabotage Linux
	echo -n " Type a number: "
	read -r choice
	choice=${choice//[^0-9]}
	#choice=${choice,,}
	(( debug > 0 )) && echo sanitized choice is $choice
	echo

	case $choice in
		1)	tpl=bullseye
			nextavailable
			askname
			sudo /root/xen/new-resource.bash $tpl $avail && \
			sudo /root/xen/newguest-debian.bash $avail $name \
				&& echo $avail,$tpl,$name >> $HOME/guests.csv
			echo " Press Enter to return to previous menu"
			read -r
			exit
			;;
		2)	tpl=netbsd-current
			nextavailable
			askname
			sudo /root/xen/new-resource.bash $tpl $avail && \
			sudo /root/xen/newguest-netbsd.bash $avail $name \
				&& echo $avail,$tpl,$name >> $HOME/guests.csv
			echo " Press Enter to return to previous menu"
			read -r
			exit
			;;
		3)	tpl=slack150
			#tpl=slackf2fs
			nextavailable
			askname
			sudo /root/xen/new-resource.bash $tpl $avail && \
			sudo /root/xen/newguest-slack.bash $avail $name \
				&& echo $avail,$tpl,$name >> $HOME/guests.csv
			echo " Press Enter to return to previous menu"
			read -r
			exit
			;;
		4)	tpl=jammy
			nextavailable
			askname
			sudo /root/xen/new-resource.bash $tpl $avail && \
			sudo /root/xen/newguest-debian.bash $avail $name \
				&& echo $avail,$tpl,$name >> $HOME/guests.csv
			echo " Press Enter to return to previous menu"
			read -r
			exit
			;;
		*)	return
			;;
	esac
}

function nextavailable {
	# grab last used drbd minor but avoid 200-254 (reserved)
	last=`cut -f1 -d, $HOME/guests.csv | sort -V | tail -1`

	if (( last < 1024 )); then
		# while starting a cluster from scratch, guest ids start with 1024 to match DNAT
		(( avail = 1024 ))
	else
		# simply increment to grab the next available guest id
		# TODO also re-use previously removed guest ids
		(( avail = last + 1 ))
	fi

	(( debug > 0 )) && echo avail is $avail

	# 64999+ are reserved
	(( avail > 65998 )) && echo cannot handle that amount of instances && exit 1
}

function askname {
	echo -n " What target system hostname would you like to set? [dnc$avail]"
	read -r input
	name=${input//[^A-Za-z0-9-]}
	name=${name,,}
	[[ -z $name ]] && name=dnc$avail
}

if (( auto == 1 )); then
	tpl=slack150
	#tpl=slackf2fs
	nextavailable
	name=dnc$avail
	sudo /root/xen/new-resource.bash $tpl $avail
	sudo /root/xen/newguest-slack.bash $avail \
		&& echo $avail,$tpl,$name >> $HOME/guests.csv
	exit
fi

clear

while true; do
	asksystem
done

echo Press Enter key to exit
read -r



manage-guests.bash
#!/bin/bash
set -e

debug=0

function bomb {
        time=`date "+%Y-%m-%d %H:%M:%S"`

        echo "$time - error: $@" >> /var/tmp/nobudget.$USER.error.log 2>&1
        echo
        echo "error: $@"
        echo

	unset time
        exit 1
}

function showhelp {
	cat <<EOF

        list            list all guests and their statuses
        new		deploy a new guest
        terminate       terminate a guest and its storage

        start		power on a guest
        reboot          reboot a guest
        stop		gracefully shutdown and power off a guest
        destroy         brutally power off a guest

        console         connect to a guest console

	exit		return to main menu

EOF
	#quit		close this session
}

function allguests {
	# TODO hmm that one wouldnt get refreshed in case of a new guest gets ordered
	[[ -z $folders ]] && folders=`find /data/guests/dnc* -type d -maxdepth 0`
	for guestpath in $folders; do
		guest=${guestpath##*/}

		echo -n $guest | sed 's/^/\t/'
		cat $guestpath/state | sed 's/^/\t/'

		unset guest
	done
	unset guestpath

	# might be re-used
	#unset folders
}

# TODO check that $node is still accurate for those functions
# see function whatnode

function cr {
	[[ -z $node ]] && bomb function cr requires \$node
	[[ -z $guest ]] && bomb function cr requires \$guest

	echo creating $guest
	ssh $node -t xl create $guest && echo created
	echo
}

function co {
	[[ -z $node ]] && bomb function co requires \$node
	[[ -z $guest ]] && bomb function co requires \$guest

}

function shu {
	[[ -z $node ]] && bomb function shu requires \$node
	[[ -z $guest ]] && bomb function shu requires \$guest

}

function des {
	[[ -z $node ]] && bomb function des requires \$node
	[[ -z $guest ]] && bomb function des requires \$guest

}

function manage_guests {
	echo -n "> "
	read -r cmd
	#cmd=${cmd//[^A-Za-z0-9-]}
	cmd=${cmd//[^A-Za-z\?]}
	cmd=${cmd,,}
	(( debug > 0 )) && echo sanitized cmd is $cmd

	# TODO also allow < > and ^D
	case $cmd in
		\?)
			showhelp
			;;
		co*)
			echo reaching out to $guest console
			ssh $node -t xl console $guest && echo TODO CLOSING CONSOLE || echo TADAAA
			echo
			;;
		cr*)
			cr
			;;
		des)
			echo destroying $guest
			ssh $node -t xl destroy $guest && echo destroyed
			echo
			;;
		l*)
			allguests
			;;
		n*)
			$HOME/nobudget/new-guest.bash
			;;
		r*)
			echo rebooting $guest
			ssh $node -t xl reboot $guest && echo rebooting
			echo
			;;
		shu)
			echo shutting down $guest
			ssh $node -t xl shutdown $guest && echo shutting down
			echo
			;;
		terminate)
			echo TODO TERMINATE A GUEST
			;;
		e*|q*)
			#echo
			#echo -n Press Enter key to return to main menu
			#read -r
			exit
			;;
		#q*)
		#	echo
		#	echo -n Press Enter to close the session
		#	read -r
		#	exit
		#	;;
		*)
			;;
	esac
	unset cmd
}

clear

[[ -z $USER ]] && bomb USER not defined - should be `whoami`
[[ -z $HOME ]] && bomb HOME not defined for $USER
[[ ! -d /data/users/ ]] && bomb /data/ does not seem to be mounted - cluster state is non-optimal
[[ ! -d $HOME/ ]] && bomb $HOME/ does not exist

[[ ! -x /usr/local/bin/new-guest.bash ]] && bomb could not find /usr/local/bin/new-guest.bash executable

echo
echo MANAGE GUESTS
echo
echo Enter ? for help

while true; do
	manage_guests
done



new-guest.bash
#!/bin/bash

(( debug = 0 ))

function bomb {
        time=`date "+%Y-%m-%d %H:%M:%S"`

        echo "$time - error: $@" >> /var/tmp/nobudget.$USER.error.log 2>&1
        echo
        echo "error: $@"
        echo

	unset time
        exit 1
}

function asksystem {
	cat <<EOF
 Which system would you like to deploy?

  1. Debian 11 (Bullseye)
  2. NetBSD current (Sep 2022)
  3. Slackware Linux 64-bit 15.0
  4. Ubuntu 22 (Jammy Jellyfish)

EOF
# Sabotage Linux

	echo -n " Type a number: "
	read -r choice
	choice=${choice//[^0-9]}
	#choice=${choice,,}
	(( debug > 0 )) && echo sanitized choice is $choice
	echo

	case $choice in
		1)	tpl=bullseye
			nextavailable
			askname
			su -c "/root/xen/new-resource.bash $tpl $avail" root && \
			su -c "/root/xen/newguest-debian.bash $avail $name" root \
				&& echo $avail,$tpl,$name >> $HOME/guests.csv \
				&& ready
			echo " Press Enter to return to previous menu"
			read -r
			exit
			;;
		2)	tpl=netbsd-current
			nextavailable
			askname
			su -c "/root/xen/new-resource.bash $tpl $avail" root && \
			su -c "/root/xen/newguest-netbsd.bash $avail $name" root \
				&& echo $avail,$tpl,$name >> $HOME/guests.csv \
				&& ready
			echo " Press Enter to return to previous menu"
			read -r
			exit
			;;
		3)	tpl=slack150
			#tpl=slackf2fs
			nextavailable
			askname
			su -c "/root/xen/new-resource.bash $tpl $avail" root && \
			su -c "/root/xen/newguest-slack.bash $avail $name" root \
				&& echo $avail,$tpl,$name >> $HOME/guests.csv \
				&& ready
			echo " Press Enter to return to previous menu"
			read -r
			exit
			;;
		4)	tpl=jammy
			nextavailable
			askname
			su -c "/root/xen/new-resource.bash $tpl $avail" root && \
			su -c "/root/xen/newguest-debian.bash $avail $name" root \
				&& echo $avail,$tpl,$name >> $HOME/guests.csv \
				&& ready
			echo " Press Enter to return to previous menu"
			read -r
			exit
			;;
		*)	return
			;;
	esac
}

function nextavailable {
	# grab last used drbd minor but avoid 200-254 (reserved)
	last=`cut -f1 -d, /data/users/*/guests.csv | sort -V | tail -1`

	# that works even though we might have an empty variable (newly installed cluster)
	if (( last < 1024 )); then
		# while starting a cluster from scratch, guest ids start with 1024 to match DNAT
		(( avail = 1024 ))
	else
		# grab the next available guest id
		(( avail = last + 1 ))
	fi

	(( debug > 0 )) && echo avail is $avail

	# 64999+ are reserved
	(( avail > 65998 )) && echo guest id 64999? - cluster is too old or cannot handle that amount of instances && exit 1
}

function askname {
	echo -n " What target system hostname would you like to set? [dnc$avail]"
	read -r input
	name=${input//[^A-Za-z0-9-]}
	name=${name,,}
	[[ -z $name ]] && name=dnc$avail
}

if (( auto == 1 )); then
	tpl=slack150
	#tpl=slackf2fs
	nextavailable
	name=dnc$avail
	su -c "/root/xen/new-resource.bash $tpl $avail" root
	su -c "/root/xen/newguest-slack.bash $avail" root \
		&& echo $avail,$tpl,$name >> $HOME/guests.csv
	exit
fi

function ready {
	cat <<EOF
 Wait 30 seconds until spanning tree settles down and you will be able reach the system as follows.

	ssh pmr.angrycow.ru -p $avail -l root

EOF
}

clear

[[ -z $USER ]] && bomb USER not defined - should be `whoami`
[[ -z $HOME ]] && bomb HOME not defined for $USER
[[ ! -d /data/users/ ]] && bomb /data/ does not seem to be mounted - cluster state is non-optimal
[[ ! -d $HOME/ ]] && bomb $HOME/ does not exist

[[ $1 = auto ]] && auto=1

echo
echo NEW GUEST
echo
while true; do
	asksystem
done

# new-guest can be called either from main menu or from manage-guests
echo -n Press Enter key to return to previous menu
read -r

